### PR TITLE
implement in memory TripModel storage

### LINKED ISSUE #8: Create Storage Interface
Create Storage Interface and In-memory Implementation.

### PR DESCRIPTION
**Pull request by cli1903**

### CODE REVIEW

File: frontend/src/test/java/com/google/sps/InMemoryTripModelStorageTest.java [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+  public static final TripModel LONDON_TRIP_DUPLICATE_ID = TripModel.builder()
+      .setId("a")
+      .setName("London Trip")
+      .setUserId("person 2")
+      .setLocationLat(32.333)
+      .setLocationLong(52.1)
+      .build();
+
+
+  /**
```
Comment (bsidhom) [inline frontend/src/test/java/com/google/sps/InMemoryTripModelStorageTest.java:56]: FYI, you don't usually need to worry about adding javadoc to tests, especially if the test is trivial and the parameters and expectations are clear from the method name (as in this case). Please do continue to document subtle test cases though.

File: frontend/src/main/java/com/google/tripmeout/frontend/TripModel.java [hunk 2]
```diff
@@ -8,6 +8,7 @@ public static Builder builder() {
     return new AutoValue_TripModel.Builder();
   }
 
+  public abstract Builder toBuilder();
```
Comment (bsidhom) [inline frontend/src/main/java/com/google/tripmeout/frontend/TripModel.java]: Nit: please move this below the TripModel fields and add a separating space. This makes it clear that toBuilder is not semantically part of this value class.

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryTripModelStorage.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class InMemoryTripModelStorage {
+  Map<String, TripModel> storage = new ConcurrentHashMap<>();
+
+  public void addTrip(TripModel trip) throws TripAlreadyExistsException {
+    synchronized(storage) {
```
Comment (bsidhom) [inline frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryTripModelStorage.java]: Nit: space after `synchronized` and before the opening `(`. The formatter _should_ cover this (I think).

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryTripModelStorage.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+import com.google.tripmeout.frontend.error.TripNotFoundException;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class InMemoryTripModelStorage {
+  Map<String, TripModel> storage = new ConcurrentHashMap<>();
+
+  public void addTrip(TripModel trip) throws TripAlreadyExistsException {
```
Comment (bsidhom) [inline frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryTripModelStorage.java:16]: Did you already talk with @KoopaKing about making our custom exceptions extend `RuntimeException`  versus `Exception`? This is fine as it is, but it's important to have considered this and it largely depends on how it's consumed in the frontends.

In particular, if the servlets will just catch all instances of `TripMeOutException` at the top-level and perform automatic exception translation (which I'd probably recommend), then you may want to consider making this a `RuntimeException`. In that case, you can drop the `throws` clauses on all methods here. That might especially make sense if there there are multiple, intermediate-stage callers who cannot/do not know how to handle the exceptions thrown here.

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryTripModelStorage.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+import com.google.tripmeout.frontend.TripModel;
+import com.google.tripmeout.frontend.error.TripAlreadyExistsException;
+import com.google.tripmeout.frontend.error.TripNotFoundException;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class InMemoryTripModelStorage {
```
Comment (bsidhom) [inline frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryTripModelStorage.java]: Please make this implement `TripStorage`. Doing so will allow this to be plugged in wherever a `TripStorage` is required. It will also reveal a compile-time error due to the exception signature not matching that of the interface. (I'd recommend using `RuntimeException`s for these anyway, unless you've already talked about this with Chris).

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryTripModelStorage.java [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+  Map<String, TripModel> storage = new ConcurrentHashMap<>();
```
Comment (bsidhom) [inline frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryTripModelStorage.java:13]: You can use a regular `HashMap` here since you're using external synchronization.

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryTripModelStorage.java [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+      if (trip == null) {
+        throw new TripNotFoundException("Trip with id: " + tripId + 
+            " not found in storage");
+      }
+      TripModel updatedTrip = trip.toBuilder()
+          .setLocationLat(latitude)
+          .setLocationLong(longitude)
+          .build();
+
+      storage.put(tripId, updatedTrip);
```
Comment (bsidhom) [inline frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryTripModelStorage.java]: You were asking earlier if you should use `ConcurrentHashMap` or `synchronized` blocks in this class. This is a use case where you'll want a `synchronized` block, since you're reading map data, performing an update, and writing it back. However, for this to safely make map updates atomic, _all_ accesses need to be synchronized in the same way (which you're already doing in this change).

As I mentioned above, since you're using synchronization this way, you can drop `ConcurrentHashMap` in favor of a regular old `HashMap` (the former incurs some thread-safety overhead that you don't need).

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryTripModelStorage.java [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+      }
+      throw new TripNotFoundException("Trip with id: " + tripId + 
+          " not found in storage");
+    } 
+  }
+
+  public List<TripModel> getAllUserTrips(String userId) {
+    List<TripModel> trips = new ArrayList<>();
+    storage.forEach((tripId, trip) -> {
+      if (userId.equals(trip.userId())) {
```
Comment (bsidhom) [inline frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryTripModelStorage.java:63]: The fact that you have to iterate through the map and check for individual matches here shows that a singly-keyed map is probably not the ideal choice here (likely why Chris recommended using a table earlier). There are options such as storing an additional map indexed by user id in order to make these lookups quick. However, that may be overkill since we only intend to use this for testing (as of now; that could change later). (Note that using a second map adds some subtlety with respect to how you do synchronization, but it's still manageable.)

More importantly, consider what behavior you want here when the underlying storage is updated while you iterate over trips here. You probably don't want a `ConcurrentModificationException` or undefined behavior. In that case, you'll need to synchronize on the map in the same way all other accesses are done. There will be performance implications since the entire map will be locked while you do this iteration, but it's the safer approach. For this test-oriented storage system, we expect the data to be small and accesses to be fast, so this probably doesn't matter.

If we did want to support fast access and safe concurrency, the best bet as of now would likely be a second map index, accessed with synchronization. (Better yet, using [persistent data structures](https://en.wikipedia.org/wiki/Persistent_data_structure) which would minimize the need for external locking.) There's also the option of [optimistic locking](https://en.wikipedia.org/wiki/Optimistic_concurrency_control), in which case you assume that concurrent modifications are infrequent and you can somehow roll back any changes and bubble an error up to the client (perhaps to be retried). In this latter case, you _might_ actually want a `ConcurrentHashMap`, which would throw an exception in the bad case without corrupting data.

Anyway, there are a lot of subtleties and design decisions here that it's good to be aware of. For now (depending on your use cases), it's probably best to wrap this in a `synchronized` block and call it a day.

**Review by bsidhom — CHANGES_REQUESTED** [review]
Please run `make format` in order to make sure everything's formatted correctly. You may need to merge upstream changes in order to make that work.

File: frontend/src/test/java/com/google/sps/InMemoryTripModelStorageTest.java [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+        storage.addTrip(LONDON_TRIP_DUPLICATE_ID));
+    
+  }
+
+  /**
+   * test that adding a different TripModel object does not cause an exception
+   * to be thrown
+   */
+  @Test
+  public void addTrip_tripNotInStorage_exceptionNotThrown()
```
Comment (bsidhom) [inline frontend/src/test/java/com/google/sps/InMemoryTripModelStorageTest.java]: This case may just be adding more test burden without being very revealing. You generally want each test case to end in an assertion (ideally the _only_ assertion made).

Here, it seems like the semantic behavior we're testing is whether the storage system correctly supports multiple trips with the same name. If so, then we should be testing that those values can actually be retrieved and match their expected values. (If that's not what you're trying to test, then figure out how to spell the test case with storage interactions that can be asserted.)

File: frontend/src/test/java/com/google/sps/InMemoryTripModelStorageTest.java [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+      .setLocationLong(52.1)
+      .build();
+
+
+  /**
+   * test that an exception is thrown when adding a TripModel that has already 
+   * been added to storage
+   */
+  @Test
+  public void addTrip_tripAlreadyAdded_throwsException()
```
Comment (bsidhom) [inline frontend/src/test/java/com/google/sps/InMemoryTripModelStorageTest.java]: Nit: it seems like the condition you want to test here is that a trip with the _same ID_ already exists. If so, I recommend updating the description to match that; you may even want to change the test case to try and store a _different_ trip model object that happens to have the same key. (I'll leave that choice up to you; doesn't seem super important here.)

File: frontend/src/test/java/com/google/sps/InMemoryTripModelStorageTest.java [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+   * object
+   */
+  @Test
+  public void addTrip_getTrip_addedTripEqualsGottenTrip() 
+      throws TripAlreadyExistsException, TripNotFoundException {
+    InMemoryTripModelStorage storage = new InMemoryTripModelStorage();
+    storage.addTrip(LONDON_TRIP1);
+    storage.addTrip(LONDON_TRIP2);
+    TripModel gottenTrip = storage.getTrip(LONDON_TRIP2.id());
+    Assert.assertEquals(LONDON_TRIP2, gottenTrip);
```
Comment (bsidhom) [inline frontend/src/test/java/com/google/sps/InMemoryTripModelStorageTest.java]: I strongly encourage you to switch to [Truth](https://truth.dev/) for these assertions. This results in more readable test code and error messages. The fluent style makes it clear which of the inputs you're testing and what the expected value is. For example, this might be translated as such:

```java
// Optionally, at top of file. Otherwise, you'd invoke it below as Truth.assertThat().
import static com.google.common.truth.Truth.assertThat;

// Within test case
assertThat(gottenTrip).isEqualTo(LONDON_TRIP2)
```

**Review by bsidhom — CHANGES_REQUESTED** [review]

File: frontend/src/test/java/com/google/sps/InMemoryTripModelStorageTest.java (hunk 10 above)
Comment (bsidhom) [reply frontend/src/test/java/com/google/sps/InMemoryTripModelStorageTest.java]: Just realized that the matching trip ID case is already handled. In that case, you can just remove this test case. In general, you want to test exactly the edge cases you care about in as few ways/places as possible to make refactoring easier if, say, the interface semantics change.

File: frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryTripModelStorage.java (hunk 8 above)
Comment (bsidhom) [reply frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryTripModelStorage.java:63]: One more point to add: if you decide to use optimistic concurrency above, you'll need to use [`ConcurrentHashMap.compute`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#compute-K-java.util.function.BiFunction-) to perform atomic read-write updates above. Lone `get` calls can be left bare.

Not suggesting you do that here, just pointing out the option.

**Review by bsidhom — APPROVED** [review]
LGTM

**Review by chuckluv — APPROVED** [review]
